Offer an option to specify a cache for conditionally fetching public keys

francisfuzz opened the pull request:
## Summary

Closes #9 

Previously, clients who fetched verification keys from the server had no way to cache those keys and reuse them for other requests.

This PR proposes a change using the GitHub API's [conditional requests](https://docs.github.com/en/rest/using-the-rest-api/best-practices-for-using-the-rest-api?apiVersion=2022-11-28#use-conditional-requests-if-appropriate) feature: clients can optionally specify a cache for their keys and only fetch new ones if the cache is outdated.

BREAKING CHANGE: `verifyRequestByKeyId() now returns an object with a `isValid` property and a`cache` property.

Before

```js
const isValid = await verifyRequestByKeyId();
```

After

```js
const { isValid, cache } = await verifyRequestByKeyId();
```

BREAKING CHANGE: `fetchVerificationKeys()` now returns an object with a `keys` property and a`cache` property.

Before

```js
const keys = await fetchVerificationKeys();
```

After

```js
const { keys, cache } = await fetchVerificationKeys();
```

francisfuzz commented on lib/verification.js:
```diff
@@ -4,36 +4,45 @@ import { createVerify } from "node:crypto";
 
-  return data.public_keys;
+  if (response.status === 304) {
+    return { cacheId: cache.id, keys: cache.keys };
```
```suggestion
    return cache;
```

francisfuzz commented on test/verification.test.js:
```diff
@@ -180,5 +180,8 @@ test("fetchVerificationKeys()", async (t) => {
+  t.deepEqual(result, { cacheId: "", keys: publicKeys });
 });
+
+// Up next: test the cache
```
⏸️ Paused here to mark progress. I don't (yet) understand how `MockAgent` works, but I'll see what a good night's sleep will do and pick this back up tomorrow or Friday. ✌️

gr2m replied:
I'm still confused by the multiple layers of network requests by undici's `fetch`. No need to understand all the inner wokrings, you should be fine copying from other tests.

gr2m commented on lib/verification.js:
```diff
@@ -4,36 +4,45 @@ import { createVerify } from "node:crypto";
 
-/** @type {import('..').VerifyRequestByKeyIdInterface} */
-export async function verifyRequest(rawBody, signature, key) {
+/** @type {import('..').VerifyRequest} */
+export async function verifyRequest(rawBody, signature, keyId) {
```
The type change is correct, but the variable name should remain `key`. This method needs the actual key, not the ID of the key

gr2m commented on index.d.ts:
```diff
@@ -27,7 +37,7 @@ interface VerifyRequestByKeyIdInterface {
     keyId: string,
     requestOptions?: RequestOptions,
-  ): Promise<boolean>;
+  ): Promise<{ isValid: boolean; cacheId: string }>;
```
I think we will have to return `{ isValid, cache }`, where `cache` is the object with both the id and the keys. Otherwise I can't get the information I need for subsequent `verifyRequestByKeyId()` calls

gr2m commented on index.d.ts:
```diff
@@ -6,6 +6,10 @@ type RequestInterface = typeof request;
 type RequestOptions = {
   request?: RequestInterface;
   token?: string;
+  cache?: {
+    id: string;
+    keys: VerificationPublicKey[];
+  };
```
as we have that signature in a multiple places, I think we could define a new type such as `VerificationKeysCache` to remove the duplication

gr2m commented on index.test-d.ts:
```diff
@@ -76,7 +77,7 @@ export async function verifyRequestTest(
 export async function fetchVerificationKeysTest() {
   const result = await fetchVerificationKeys();
-  expectType<{ cacheId: string; keys: VerificationPublicKey[] }>(result);
+  expectType<VerificationKeysCache>(result);
```
for tests, I'd leave the explicit assertions. This will help us to catch unintentional API changes. Say we accidentally remove or change a key in the `VerificationKeysCache` type. This test would still pass. But having it being explicit will fail, helping us prevent a regression.

It's also a good reminder to add additions to changelogs, as each addition needs to be added to the assertion

francisfuzz commented on lib/verification.js:
```diff
@@ -23,17 +23,35 @@ export async function verifyRequest(rawBody, signature, key) {
+    const cacheId = response.headers.etag || "";
+    return { id: cacheId, keys: response.data.public_keys };
+  } catch (error) {
+    if (error.status === 304) {
```
When I originally [wrote up the tests](https://github.com/copilot-extensions/preview-sdk.js/pull/81/files#diff-888235d23d4180ab42cebcaa7b47a7d3ec586a7290a7e5acaf4d05fe5ca9eb04R188-R237), I thought this would "just work": however, I saw that it failed because an error was thrown.

Re-reviewing my work, the only place where an error would be thrown would be on the request for the public keys](https://github.com/copilot-extensions/preview-sdk.js/pull/81/files#diff-7b02190766e81ebccdc674c1257d3ceead3e9c97c6d0968a5856749625cb13f3R37).

I researched `octokit/request.js` and looked into [what happens when a `status: 304` is returned](https://github.com/octokit/request.js/blob/bec87c90d08449815ed8cac4f86eb3d082ed5f9c/src/fetch-wrapper.ts#L124-L131): it turns out that it captures the data, but seems to go on to throw a new `RequestError` with the full request-response pair. 😅 

I'm sure there's a reason (that I'm unaware of) for why it's written in this way, so I chose to catch the error and check its status; should it be a `304`, we go on to return the `cache`. Otherwise (for say a `4xx` or `5xx` error), we'll throw the error. 

Happy to make updates to the error handling here with some more guidance! 😉 👼

gr2m replied:
`@octokit/request` is throwing an error in case of a `304` because it cannot give you the response body. Throwing an error was the only way that I could think of to give developers a way to handle the case when a conditional request matches, meaning the response is 304 without any body. In that case it's up to the developer to provide the body instead.

I feel a proper high-level API for conditional request is really missing from the Octokit SDK, I never got around to give it more thought. 

All this to say, the way you implemented it is the correct way.

francisfuzz commented:
@gr2m - Thank you for all of your preliminary feedback in draft 🙇 

I wrote the last test and [left this comment about it](https://github.com/copilot-extensions/preview-sdk.js/pull/81#discussion_r1757643467) for awareness. Happy to make any changes to get this over the line, especially around the way I've defined the types and the tests! 👂

gr2m commented on index.d.ts:
```diff
@@ -14,11 +19,11 @@ export type VerificationPublicKey = {
 
 interface VerifyRequestInterface {
-  (rawBody: string, signature: string, key: string): Promise<boolean>;
+  (rawBody: string, signature: string, key: string): boolean;
```
The `verifyRequset` method is async, we need to keep the `Promise<...>`
```suggestion
  (rawBody: string, signature: string, key: string): Promise<boolean>;
```

francisfuzz replied:
Meant to follow up on this: I originally walked through the block of code and, originally saw that the operations were all synchronous (assertions; createVerify; update; verify). However, seeing the prepended `async` on the `.verifyRequest` method helped me see why we should be returning a `Promise<boolean>` there -- thanks for the correction @gr2m!

gr2m commented on lib/parse.js:
```diff
@@ -22,15 +22,15 @@ export function transformPayloadForOpenAICompatibility(payload) {
 
   return {
-    isValidRequest,
+    isValidRequest: isValid,
```
The `verifyAndParseRequest` method should also accept `options.cache` and return a `cache` key. It's the most high-level of the methods that probably most users end up using right now. Without exposing these, they won't be able to implement key caching.

francisfuzz replied:
Thanks @gr2m, I'll get to work on this 👁️

francisfuzz replied:
> The verifyAndParseRequest method should also accept options.cache and return a cache key. It's the most high-level of the methods that probably most users end up using right now. Without exposing these, they won't be able to implement key caching.

@gr2m - https://github.com/copilot-extensions/preview-sdk.js/pull/81/commits/81a6f3387ba94444b936223123f39eafa23835c2 is my interpretation of this --when you have cycles, may I please ask for your review in that's what you envisioned? Happy to correct anything that I may have missed or errantly understood! 🙇

gr2m replied:
looks good to me now, thank you 👍🏼

gr2m commented on test/verification.test.js:
```diff
@@ -34,6 +34,14 @@ test("verifyRequestByKeyId()", async (t) => {
     return fetch(url, opts);
   }
 
+  const publicKeys = [
+    {
+      key: CURRENT_PUBLIC_KEY,
+      key_identifier: KEY_ID,
+      is_current: true,
+    },
+  ];
```
I'd move it below `SIGNATURE` and use `PUBLIC_KEYS`, as the value does not change in-between tests.

francisfuzz replied:
Done in https://github.com/copilot-extensions/preview-sdk.js/commit/482816832601ff055f5c43f3451a0928a617abc7 ✅

francisfuzz commented:
@gr2m - Question on scope for documentation: should I/we update the PR to include the new "state" of things since this is a breaking change? Or, do you prefer documentation-specific changes to land separately in their own PR? 💭

gr2m commented:
> @gr2m - Question on scope for documentation: should I/we update the PR to include the new "state" of things since this is a breaking change? Or, do you prefer documentation-specific changes to land separately in their own PR

very good call out. As we are changing behavior, we have to update the README.md file where the current API is documented as part of this pull request

francisfuzz commented:
I can get on it Monday!

francisfuzz commented:
Leaving an update: I've updated the `README.md` with most of what's changed. I'll have a final look over when I'm back online. ✌️

gr2m commented on index.test-d.ts:
Can we add an example that test `verifyRequestByKeyId()` with the cache argument?

francisfuzz replied:
Resolved in https://github.com/copilot-extensions/preview-sdk.js/pull/81/commits/c7b600eb9f1980800c5c647c9b9cac3a64c9ca7b

gr2m commented on lib/verification.js:
```diff
@@ -23,17 +23,34 @@ export async function verifyRequest(rawBody, signature, key) {
-  });
+      : {};
+
+    if (cache.id) headers["If-None-Match"] = cache.id;
```
Did you try if the changes work? I found the `etag` response and `If-None-Match` request header to be tricky in the past, there is some prefix that needs to be encoded ... just want to make sure we test the code in real-life before we merge the PR. If you run into any trouble let me know I can look what I did in other projects

francisfuzz replied:
I was _just_ thinking about test running the changes. Will report back! 🫡

francisfuzz replied:
Example script:

```js
import { request } from "@octokit/request";
import { fetchVerificationKeys } from "./index.js";

async function main() {
    const token = process.env.GITHUB_TOKEN;
    if (!token) throw new Error("GITHUB_TOKEN is not defined");

    const r = request.defaults({
        headers: {
            Authorization: `token ${token}`,
        }
    });

    const { id, keys } = await fetchVerificationKeys({ token });
    const cache = { id, keys };

    const firstRateLimitCheck = await r("GET /rate_limit");
    const firstRateLimitRemaining = firstRateLimitCheck.data.resources.core.remaining;
    console.log("Rate limit after fetching keys:", firstRateLimitRemaining);

    const response = await fetchVerificationKeys({ token, cache });

    const secondRateLimitCheck = await r("GET /rate_limit");
    const secondRateLimitRemaining = secondRateLimitCheck.data.resources.core.remaining;
    console.log("Rate limit after fetching keys from cache:", secondRateLimitRemaining);
}

main().catch(console.error);
```

Running it:

```js
@francisfuzz ➜ /workspaces/preview-sdk.js (9-cache) $ node example.js 
Rate limit after fetching keys: 4953
Rate limit after fetching keys from cache: 4953
```

gr2m commented on test/verification.test.js:
```diff
@@ -180,5 +169,44 @@ test("fetchVerificationKeys()", async (t) => {
+    cache,
+  });
+
+  t.deepEqual(result, cache);
```
can we add another test which sends the `fetchVerificationKeys()` request twice to simulate real-life behavior? The first one should respond with 200 and the body and a `etag` header. And the 2nd responds with a 304. 

I'm not sure how undici is working with mocking, defining different replies for the same request can be tricky. We can also file a follow up issue and take care of it later

francisfuzz replied:
Took a stab at this in https://github.com/copilot-extensions/preview-sdk.js/pull/81/commits/fb4f95c46074eb97680f027ee7290b6513b065b4

francisfuzz replied:
Let me update the `etag` to be something that looks like what we'd expect from the GitHub API - please hold!

francisfuzz commented:
@gr2m - This is ready for re-review! I've updated the README, incorporated your suggested feedback in today's most recent review, and re-ran the tests locally to ensure everything passes. I appreciate your time and can make any changes as needed to get this out (or, hold until _after_ tomorrow's stream 😉 ).

gr2m commented:
I don't think this is working yet.

Here is a test script

```js
// .gregor/conditional-requests.js
import { request } from "@octokit/request";

import { fetchVerificationKeys } from "../index.js";

let counter = 0;

const requestWithLog = request.defaults({
  request: {
    hook(request, options) {
      counter += 1;
      return request(options);
    },
  },
});

const result1 = await fetchVerificationKeys({ request: requestWithLog });
const result2 = await fetchVerificationKeys({ request: requestWithLog });

console.log({ counter });
```

This script currently logs out 

```
{ counter: 2 }
```

But it should log out

```
{ counter: 1 }
```

gr2m commented:
Ignore me, my script was incorrect, this one is correct

```js
import { request } from "@octokit/request";

import { fetchVerificationKeys } from "../index.js";

const requestWithLog = request.defaults({
  request: {
    async hook(request, options) {
      try {
        const response = await request(options);
        console.log("no cache hit");
        return response;
      } catch (error) {
        if (error.status === 304) {
          console.log("cache hit");
          throw error;
        }

        console.log("unexpected error");
        throw error;
      }
    },
  },
});

const cache = await fetchVerificationKeys({ request: requestWithLog });
await fetchVerificationKeys({ request: requestWithLog, cache });
```

And the output is as expected

```
no cache hit
cache hit
```

gr2m commented on lib/verification.js:
```diff
@@ -23,51 +23,71 @@ export async function verifyRequest(rawBody, signature, key) {
+    });
 
-  return data.public_keys;
+    if (response.status === 304) {
+      return cache;
+    }
+
```
The code will never get here, Octokit is throwing an error in case of a `304`.

```suggestion
```

gr2m reviewed (approved):
Great work on this one! We support caching now yay!
